kernel/model/model.go

```diff
@@ -49,8 +50,8 @@ type Model struct {
 	Parent *Model
 
 	Scope
-	Regions Regions
-
+	Regions             Regions
+	ModelFactories      []Factory // Factories that change the model structure, eg: add/remove hosts
```

Comment:
Why would this be different than a `Factory`? Why add another `Factory` concept?

Reply:
It's to separate Factories which mutate model structure vs ones that don't. The structure manipulation has to be done before we bind the label and environment, otherwise things like IP addresses don't end up in the right spot, b/c their intended targets aren't there yet. Other Factories get done after the label/binding are applied, and it didn't feel right to move them all in front of label/env binding. 

Let me know if that makes sense, if you've got ideas on how that could be solved differently.

Reply:
I don't think by nature factories need to be inserted into the model in the correct order. What you're doing with "model" factories is no different.. they just happen first. Not sure that it's a useful distinction. It's more powerful to just have "factories" and put them in the correct order.

Reply:
Are you arguing there should be just one set of factories and that they should all be invoked before the label and environment is bound? I suspect that will break other factories which rely on variables to be set.

Reply:
Why can't all of the factories run _after_ the label and environment are bound? The label and environment binding just feed static data to the factories and the model?

Reply:
Maybe label/environment binding gets moved into a `Factory` and then the ordering can be controlled however we want per-model?

Reply:
Moving label/env binding into a factory would work. The templating stuff really does need to run before binding, otherwise the binding doesn't have anything to apply values to.

Reply:
We should probably loop this into our conversation tomorrow.